### ISSUE TITLE
Use a modern fast hash

### ISSUE THREAD

**Issue by solardiz** [openwall/john#4350]
We might want to get a modern fast hash function like t1ha into the tree and use it throughout the codebase, including in "unique". The current 2-characters-at-a-time hash function that we use in a few places including "unique" is probably not the fastest we can have with decent properties on modern CPUs. https://github.com/erthink/t1ha

Stretch goal: use a keyed hash with a key randomly generated at startup to make JtR immune to hashDoS. Would probably have slight performance impact at least on reading the key from a global variable each time a hash is computed?

**Comment by ghost** [comment]
I've been experimenting with these ideas. Below, what I found:

Reference value, from upstream:
```
rm -f original.txt; time /snap/john-the-ripper/current/run/unique -v -i=500k.in original.txt

Hash size 25 (32 Mi/256 MiB), input buffer 2 GiB. Total alloc. 2304 MiB
Total lines read: 900000, unique lines written: 899997 (99%), no slow passes

real	0m0,305s
user	0m0,205s
sys	0m0,101s
```
*****
The values below are not from upstream source code and it is not C.

t1ha (good).
```
rm -f novo.txt; time projects/unique/target/release/unique    -v -i=500k.in novo.txt

Hash size 25 (32Mi/256MiB), input buffer 2GiB. Total alloc. 2304MiB
Total lines read: 900000, unique lines written: 899997 (99%), no slow passes.

real	0m0,349s
user	0m0,321s
sys	0m0,029s
```

WyHash (the best).
```
rm -f novo.txt; time projects/unique/target/release/unique    -v -i=500k.in novo.txt

Hash size 25 (32Mi/256MiB), input buffer 2GiB. Total alloc. 2304MiB
Total lines read: 900000, unique lines written: 899997 (99%), no slow passes.

real	0m0,329s
user	0m0,276s
sys	0m0,053s
```

Also, Highway hash (50ms - 100ms slower).

**Comment by solardiz** [comment]
Thanks, @claudioandre-br! In other words, per these quick tests our current code is the fastest? That's kind of cool, although I'd use a longer-running benchmark and rerun it a few times.

BTW, such testing covers two aspects at once: speed of hash computation and rate of hash collisions. It is possible, for example, that one of those alternative hashes had fewer collisions (produced a more uniform distribution of hash values), but was so much slower to compute that the overall effect is still negative.

**Comment by ghost** [comment]
> per these quick tests our current code is the fastest?

It's a little bit more complicated, since we are not comparing oranges and oranges. I'm using a HashSet type and (the type itself) is loosing time and memory doing unnecessary tasks. I'm going to change it to something better like a HashMap type.

> I'd use a longer-running benchmark

I'm still struggling with the upstream 'unique 'source code to understand what it's doing when the buffer size is, e.g., 10 units in size but the input file requires 30 units in size (e.g.). When my source code handles huge files, I can test this better.

**Comment by solardiz** [comment]
Oh, you wrote your own alternative to `unique`? I thought you were merely patching the different hash functions into `unique`.

**Comment by ghost** [comment]
> Oh, you wrote your own alternative to unique?

Yes, As I stated above:

> The values below are not from upstream source code and it is not C.

I'm confirming WyHash is really fast (as people says here and there) and I'm also trying to experiment using (what seems to be) the new low level reference programming language.

**Comment by ghost** [comment]
For reference:
![image](https://user-images.githubusercontent.com/1702923/195880541-e73340c5-171b-4230-abc9-40469730923c.png)

**Comment by ghost** [comment]
For reference (sorry for the spam).

Using `t1ha::t1ha2_atonce;`
```
$ rm -f novo.txt; time projects/unique/target/release/unique    -v -i=big.in novo.txt
Hash size 25 (32Mi/256MiB), input buffer 2GiB. Total alloc. 2304MiB
Total lines read: 207141975, unique lines written: 201460217 (97%), no slow passes.

real	1m30,245s
user	1m22,173s
sys	0m3,799s

```
```
$ ps ux | grep unique
claudio   729131  100 23.3 2363516 2361660 pts/0 R+   18:04   0:11 projects/unique/target/release/unique -v -i=big.in novo.txt
```

Versus

```
$ rm -f original.txt; time /snap/john-the-ripper/current/run/unique -v -hash-size=25 -i=big.in original.txt
Hash size 25 (32 Mi/256 MiB), input buffer 2 GiB. Total alloc. 2304 MiB
Slow pass 1; Total lines read: 121310709, unique lines written: 118903521 (98%)
Total lines read: 207141975, unique lines written: 201460217 (97%), 1 slow passes

real	3m27,688s
user	3m22,344s
sys	0m3,842s

```

```
$ ps ux | grep john
claudio   729919 99.1 23.3 2396620 2364864 pts/0 R+   18:08   1:52 /snap/john-the-ripper/current/run/unique -v -hash-size=25 -i=big.in original.txt
```

```
$ cmp original.txt novo.txt  # both are producing the same result
me:~/bin$

```

**Comment by solardiz** [comment]
@claudioandre-br Cool. It looks like you found a way to avoid the slow pass for the same input size in the same memory size? I wonder how you did that, given this test doesn't appear to be close to the threshold for needing a second pass. You could want to open a separate issue for proposing whatever algorithm improvement you came up with. My best guess is the average input line length in this test is (unusually?) high and you're not storing the actual strings in memory (store only hashes and offsets) or store only differences from previous lines (quick and effective compression when the input list is largely sorted). Also, we could use fewer bytes for the offsets (right now, on 64-bit we store them as 64-bit values, which is wasteful).

**Comment by solardiz** [comment]
@claudioandre-br BTW, there are multi-threaded programs like this - https://github.com/Cynosureprime/rling and https://github.com/nil0x42/duplicut - but IIRC when forced to use a single thread they performed similarly to our `unique` in my tests.

**Comment by ghost** [comment]
Well, as I was working on this I was also thinking how the new algorithm should be:
1. considering the properties of these "modern' hashes (t1ha and wyhash);
2. considering the nature of the input file (how this file can be to be usabe to crack any modern password scheme);
3. and do on;

I created one implementation that is **not** failproof.

Thinking about a modernization approach in general (including easy multi-threading in rust) as next steps.

That said, I'm not ready to solve all problems that currently upstream unique can do. It is not a fair comparative.

**Comment by solardiz** [comment]
FWIW, there's now a fork/successor to wyhash: https://github.com/Nicoshev/rapidhash

**Comment by AlekseyCherepanov** [comment]
I'll add my 2 cents: `rling` lists `xxhash` on its list of technologies used. Also hashcat's package in Debian has xxhash in dependencies.

I've read readme [here](https://github.com/Cyan4973/xxHash). 4 years ago, faster `xxhash3` was added. In benchmarks there, `xxhash3` is the fastest (`wuhash` and `rapidhash` are not listed). Also they used a framework to measure quality of hashing (e.g. collisions are mentioned).

**Comment by AlekseyCherepanov** [comment]
Ouch, `wyhash`, not `wuhash`. And it is listed in the results of tests for collisions [here](https://github.com/Cyan4973/xxHash/wiki/Collision-ratio-comparison#collision-study).

**Comment by solardiz** [comment]
I think for NUL-terminated very short strings, it'd be hard to do much better than our current hash in `unique.c` and another in `suppressor.c` (the latter produces two outputs).

So the general/third-party fast hashes we're discussing here are probably only relevant to us in cases where the length is (or can reasonably be) known in advance (already in an integer variable) and it's not very small.

**Comment by AlekseyCherepanov** [comment]
I wrote a benchmark to measure speed.

A few points discovered on this way:
- `__attribute__((noinline))` is needed to prevent inlining into the main loop.
- I used `#define B asm volatile(" nop; ");` to prevent reordering of code and make some markers to look for in `objdump -d`. It is quite convenient. 2 nops are good too.
- I used `gettimeofday()` as a recheck method because it is in vdso and does not slowdown things (though for big wordlist real syscall would not be noticeable too).
- `perf` with default settings cannot account children calls, so it would show 20x difference. `--call-graph dwarf` is needed as described [here in "Update" part](https://stackoverflow.com/questions/27742462/understanding-linux-perf-report-output).
- I expected `memchr` to be like `strlen` that knows length and can use vector instructions. `memchr` turned out slower. (`strnlen` seems even slower.)
- Long tests with `perf` can loose data. `perf` would report about it.

Overall result: xxhash seems faster. But I cannot interpret results fully: `key_hash` computes 2 hashes in parallel giving 96-bits. `xxh3_64+strlen` gives 64-bits and is ~1.5 times faster. But I used `key_hash` with index. `xxh3_64` with index is 2 times faster. For sure, `xxh3_128+strlen` is faster slightly than both `key_hash` and `line_hash`, but that assumes I use them correctly. Maybe I use `key_hash` and `line_hash` incorrectly.

Also xxhash might or might not be so fast on 32-bit cpu or on other architecture (I saw mentions that it uses handwritten vector code).

(cpu: i7 920 (64-bit, sse4.2, no avx); Linux 6.1.0, glibc 2.36, gcc 12.2.0)

Results (sorted by time descending):
```
Times (us, wall clock time including some context switches):
3439560 test1  key_hash from suppressor, dual
3011029 test2  line_hash from unique
2720522 test6  xxh3_128 + strlen
2608000 test5  xxh3_64 + memchr
2200828 test4  xxh3_64 + strlen
1685916 test3  xxh3_64 + index

Relevant part of stats from perf:
               main
               |
               |--20.79%--test1
               |
               |--18.21%--test2
               |
               |--16.44%--test6
               |          |
               |          |--12.13%--XXH3_128bits_withSeed
               |          |
               |           --1.98%--__strlen_sse2
               |
               |--15.84%--test5
               |          |
               |          |--8.96%--XXH3_64bits_withSeed
               |          |
               |           --3.99%--__memchr_sse2
               |
               |--13.30%--test4
               |          |
               |          |--8.79%--XXH3_64bits_withSeed
               |          |
               |           --2.10%--__strlen_sse2
               |
                --10.21%--test3
                          |
                          |--8.12%--XXH3_64bits_withSeed
                          |
                           --0.61%--XXH3_64bits_withSeed@plt
```

The command:
```
$ gcc -O3 test_hash.c -lxxhash && \
  perf record -e cycles -g --call-graph dwarf -- ./a.out rockyou.txt && \
  perf report --stdio
```

The code:
```c
/* Copyright is listed per-part. See below. */

#include <stdio.h>
#include <string.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <assert.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/time.h>

#include <xxhash.h>
/* add -lxxhash for linking */

/* ---------------------------------------------------------------------- */
/* Borrowed from src/common.h */
/*
 * This file is part of John the Ripper password cracker,
 * Copyright (c) 1996-99,2005,2009,2011,2013,2015 by Solar Designer
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 *
 * There's ABSOLUTELY NO WARRANTY, express or implied.
 */

#ifdef __GNUC__
#if __GNUC__ >= 5
#define MAYBE_INLINE __attribute__((gnu_inline)) __attribute__((always_inline)) inline
#elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 7) || defined(__INTEL_COMPILER) || __NVCC__
#define MAYBE_INLINE __attribute__((always_inline)) inline
#elif __GNUC__ > 3 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 1)
#define MAYBE_INLINE __attribute__((always_inline))
#else
#define MAYBE_INLINE __inline__
#endif
#elif __STDC_VERSION__ >= 199901L
#define MAYBE_INLINE inline
#else
#define MAYBE_INLINE
#endif

/* ---------------------------------------------------------------------- */
/* Borrowed from src/suppressor.c */
/*
 * This file is part of John the Ripper password cracker,
 * Copyright (c) 2022 by Solar Designer
 */

static MAYBE_INLINE uint32_t key_hash(const char *key, uint64_t *hash2)
{
	const unsigned char *p = (const unsigned char *)key;
	unsigned char *q = (unsigned char *)hash2;
	uint32_t hash1, extra;

	*hash2 = 1;
	hash1 = extra = 0;

	while (*p) {
		if (q >= (unsigned char *)hash2 + 8) {
			q = (unsigned char *)hash2;
			*hash2 += ((uint64_t)hash1 << 32) | extra;
		}
		q[0] ^= p[0];
		hash1 += p[0];
		hash1 *= 0x5a827999;
		if (!p[1])
			break;
		q[1] ^= p[1];
		extra += p[1];
		extra *= 0x6ed9eba1;
		p += 2; q += 2;
	}

	if (p - (const unsigned char *)key + !!*p > 8) /* different hash than for any short 7-bit */
		((unsigned char *)hash2)[7] |= 0x80;

	hash1 ^= extra;

	return hash1;
}

/* ---------------------------------------------------------------------- */
/* Borrowed from src/unique.c */
/*
 * This file is part of John the Ripper password cracker,
 * Copyright (c) 1998,1999,2002,2003,2005,2006,2011,2020 by Solar Designer
 * Copyright (c) 2011 by Jim Fougeron
 * Copyright (c) 2016-2019 by magnum
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 *
 * There's ABSOLUTELY NO WARRANTY, express or implied.
 */

/* from src/params.h */
#define LINE_BUFFER_SIZE		0x400
#define UNIQUE_HASH_LOG			25
#define UNIQUE_HASH_SIZE		(1 << UNIQUE_HASH_LOG)

typedef size_t uq_hash;

static int mlc;

#define HASH_MAX			0xF800000000000000ull

static size_t unique_hash_size = UNIQUE_HASH_SIZE;
static size_t unique_hash_mask = UNIQUE_HASH_SIZE - 1;

static size_t unique_hash_log = 32;
static size_t unique_hash_log_half = 16;


static uq_hash line_hash(char *line)
{
	uq_hash hash, extra;
	char *p;
	char *e = &line[mlc ? mlc : LINE_BUFFER_SIZE];

	p = line + 2;
	hash = (uint8_t)line[0];
	if (!hash)
		goto out;
	extra = (uint8_t)line[1];
	if (!extra)
		goto out;

	while (*p && p < e) {
		hash <<= 5;
		hash += (uint8_t)p[0];
		if (!p[1] || &p[1] >= e)
			break;
		extra *= hash | 1812433253;
		extra += (uint8_t)p[1];
		p += 2;
		if (hash & HASH_MAX) {
			hash ^= hash >> unique_hash_log;
			extra ^= extra >> unique_hash_log;
			hash &= unique_hash_mask;
		}
	}

	hash -= extra;
	hash ^= extra << unique_hash_log_half;

	hash ^= hash >> unique_hash_log;

	hash &= unique_hash_mask;
out:
	return hash;
}

/* ---------------------------------------------------------------------- */
/*
 * Copyright (c) 2024 by Aleksey Cherepanov
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 */

struct all {
    unsigned char *buf;
    struct {
        unsigned char *line;
        size_t length;
    } *index;
    size_t lines_k;
    size_t real_size;
    size_t fake_size;
};

__attribute__((noinline))
uint32_t test1(struct all *a)
{
    uint32_t h = 0;
    uint64_t h2tmp;
    for (int i = 0; i < a->lines_k; i++) {
        h ^= key_hash(a->index[i].line, &h2tmp);
        h ^= h2tmp;
    }
    return h;
}

__attribute__((noinline))
uq_hash test2(struct all *a)
{
    uq_hash h = 0;
    for (int i = 0; i < a->lines_k; i++) {
        mlc = a->index[i].length;
        h ^= line_hash(a->index[i].line);
    }
    return h;
}

#define xxh_seed (6564913181047158264ull)

__attribute__((noinline))
XXH64_hash_t test3(struct all *a)
{
    XXH64_hash_t h = 0;
    for (int i = 0; i < a->lines_k; i++) {
        h ^= XXH3_64bits_withSeed(a->index[i].line, a->index[i].length, xxh_seed);
    }
    return h;
}

__attribute__((noinline))
XXH64_hash_t test4(struct all *a)
{
    XXH64_hash_t h = 0;
    unsigned char *p = a->buf;
    size_t i = 0;
    while (i < a->real_size) {
        size_t len = strlen(p);
        h ^= XXH3_64bits_withSeed(p, len, xxh_seed);
        p += len + 1;
        i += len + 1;
    }
    return h;
}

__attribute__((noinline))
XXH64_hash_t test5(struct all *a)
{
    XXH64_hash_t h = 0;
    unsigned char *p = a->buf;
    size_t i = 0;
    while (i < a->real_size) {
        /* In theory, it is like strlen but can overread. Slower in practice. */
        unsigned char *e = memchr(p, '\0', a->fake_size - i);
        size_t len = e - p;
        /* size_t len = strnlen(p, a->fake_size - i); */
        h ^= XXH3_64bits_withSeed(p, len, xxh_seed);
        p += len + 1;
        i += len + 1;
    }
    return h;
}

__attribute__((noinline))
XXH64_hash_t test6(struct all *a)
{
    XXH64_hash_t h = 0;
    unsigned char *p = a->buf;
    size_t i = 0;
    while (i < a->real_size) {
        size_t len = strlen(p);
        XXH128_hash_t r = XXH3_128bits_withSeed(p, len, xxh_seed);
        h ^= r.low64;
        h ^= r.high64;
        p += len + 1;
        i += len + 1;
    }
    return h;
}

/* assert-like */
#define A(cond, ...) \
    do { \
        if (!(cond)) { \
            printf("Failure: "); \
            printf(__VA_ARGS__); \
            printf("\n"); \
            exit(1); \
        } \
    } while (0)

typedef unsigned long long u64;

int main(int argc, char *argv[])
{
    A(argc == 2,
      "Usage: %s file_to_read_and_hash", argv[0]);

    int fd = open(argv[1], O_RDONLY);
    A(fd > 0,
      "could not open file.");

    struct stat sb;
    int ret = fstat(fd, &sb);
    A(ret == 0,
      "could not fstat().");
    A((sb.st_mode & S_IFMT) == S_IFREG,
      "need a regular file.");

    /* Generously add space for reading after real data to allow vectorized access. */
#define OVERREAD_SIZE (256 + LINE_BUFFER_SIZE)

    /* We'll use it for final \0 lacking \n. */
    assert(OVERREAD_SIZE > 0);

    A(sb.st_size < SIZE_MAX - OVERREAD_SIZE,
      "file is too big.");

    unsigned char *buf = malloc(sb.st_size + OVERREAD_SIZE);
    A(buf,
      "malloc() failed for size %zu.", sb.st_size);

    memset(buf + sb.st_size, 0, OVERREAD_SIZE);

    ret = read(fd, buf, sb.st_size);
    A(ret == sb.st_size,
      "could not read whole file.");

    size_t lines_k = (buf[sb.st_size - 1] != '\n');
    for (int i = 0; i < sb.st_size; i++) {
        if (buf[i] == '\n') {
            buf[i] = '\0';
            lines_k++;
        }
    }
    printf("Loaded file: %s\nNumber of lines: %zu\n", argv[1], lines_k);

    struct all a = {
        .buf = buf,
        .index = NULL,
        .lines_k = lines_k,
        .real_size = sb.st_size,
        .fake_size = sb.st_size + OVERREAD_SIZE
    };

    a.index = calloc(lines_k, sizeof(a.index[0]));
    A(a.index,
      "could not calloc() to allocate index.");

    int after_line_end = 1;
    size_t k = 0;
    for (int i = 0; i < sb.st_size; i++) {
        if (after_line_end) {
            a.index[k].line = &buf[i];
            a.index[k].length = strlen(a.index[k].line);
            k++;
            after_line_end = 0;
        }
        if (buf[i] == '\0')
            after_line_end = 1;
    }
    assert(k == lines_k);

    printf("go\n");

    size_t result;
    u64 time_sum[7] = {};
    struct timeval time_start, time_end;

#define compute_diff_usec(end, start) \
    (((end).tv_sec - (start).tv_sec) * 1000000ull + \
     (end).tv_usec - (start).tv_usec)

/* gettimeofday() is likely to be in vdso. It is very fast. But it is
 * wall clock time, so difference between end and start times would
 * include time spent in other processes due to context switches. */

#define B asm volatile(" nop; ");

#define wrap(n, func) \
    B; \
    gettimeofday(&time_start, NULL); \
    B; \
    result = func(&a); \
    B; \
    gettimeofday(&time_end, NULL); \
    B; \
    time_sum[n] += compute_diff_usec(time_end, time_start); \
    /* printf(#func ": %zu\n", result); */ \
    B;

    for (int i = 0; i < 10; i++) {
        wrap(1, test1);
        wrap(2, test2);
        wrap(3, test3);
        wrap(4, test4);
        wrap(5, test5);
        wrap(6, test6);
    }

    printf("Times (us, include context switches):\n");
    printf("%llu test1  key_hash from suppressor, dual\n", time_sum[1]);
    printf("%llu test2  line_hash from unique\n", time_sum[2]);
    printf("%llu test3  xxh3_64 + index\n", time_sum[3]);
    printf("%llu test4  xxh3_64 + strlen\n", time_sum[4]);
    printf("%llu test5  xxh3_64 + memchr\n", time_sum[5]);
    printf("%llu test6  xxh3_128 + strlen\n", time_sum[6]);

    return 0;
}
```
